### DISCUSSION TITLE
Document withRouter: TypeError: Cannot read properties of null (reading 'basePath')

### DISCUSSION THREAD

**Discussion by erlangparasu** [vercel/next.js#36445] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Thu Apr 22 16:18:59 MSK 2021
    Binaries:
      Node: 16.14.2
      npm: 8.7.0
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 12.1.5
      react: 17.0.2
      react-dom: 17.0.2

### Describe the Bug

this.props.router is null

### Expected Behavior

this.props.router is not null

### To Reproduce

https://nextjs.org/docs/api-reference/next/router#typescript

```typescript
// ...
import { withRouter, NextRouter } from 'next/router'

class MyDocument extends Document
// ...

  render() {
  <link rel="test" href={this.props.router.basePath} />

// ...
export default withRouter(MyDocument);
```

**Comment by vadymshymko** [comment]
In your case, I think it's better to use a custom _app along with next/head and next/router instead of a custom document, because _document only works on the server side, not the client side (so any next updates won't update your href attribute of the tag links). I think you should try do something like this:
`// your custom _app.js:`
```
import Head from 'next/head';
import { useRouter } from 'next/router';

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  return (
    <div>
     <Head>
         <link rel="test" href={router.basePath} />
     </Head>
      <Component {...pageProps} />
    </div>
  )
}

export default MyApp
```
Docs about [custom document](https://nextjs.org/docs/advanced-features/custom-document), [custom app](https://nextjs.org/docs/advanced-features/custom-app), [next/router](https://nextjs.org/docs/api-reference/next/router), [next/head](https://nextjs.org/docs/api-reference/next/head)
